Move files more generic

Hwesta opened the pull request:
Rather than every Space knowing how to move to every other Space, they only need to know how to move to and from a staging area in the storage service.  The staging area must be locally accessible to the storage service (ie with tools like mv and os.rename), and is configured for each Space.

Each space implements functions to move files/folders to and from the storage service. Moving to the storage service from the original location deposits them in a staging area associated with the destination Space.  Moving from the storage service to the final destination moves the file from the destination Space's staging area to the destination path.

Also added support for moving folders, not just files.  Rsync uses -r, and extract files looks for the file on disk before trying to extract it from a Package.

No Archivematica changes needed.

Hwesta committed:
Accept AICs anywhere that AIPs work

refs #5803

Added an AIC package type.  Create, extract, delete and download AIP also
accepts an AIC. Renamed a 'file' variable to package to not conflict with
builtin. Removed temp dir once finished with it.

Commit:
Send paths as base64 strings

Paths are basically bytestrings to the OS, and we've caused ourselves
a lot of grief by trying to treat them as sets of characters instead
of bytes. Legacy data can, and frequently does, contain data in non-
Unicode encodings, mixed encodings, invalid encodings, etc.

v2 of the API sidesteps this by treating all paths as sets
of arbitrary bytes, and communicates them to/from outside services in
base64-encoded strings. The v1 API remains available and returns paths
as strings like before.

refs #6067

Commit:
File explorer: base64 decode directory names

This updates the file explorer to use the new v2 API, which requires
a new base64 JavaScript library. The library used comes from here:
https://code.google.com/p/javascriptbase64/

refs #6067

Hwesta committed:
Add and display staging path

refs #6248

Staging path will be used in future to store files being transferred
from one Space to another.

Hwesta committed:
Make moving files more generic

refs #6248

Add functions to move files/folders to and from the storage service.
Moving to the storage service from the original location deposits them
in a staging area associated with the destination Space.  Moving from
the storage service to the final destination moves the file from the
destination Space's staging area to the destination path.  The staging
area is locally accessible to the storage service, which makes the move
to/from storage service functions much simpler than trying to move to
any protocol space.

Each protocol space implements these functions, and the parent Space
calls them as needed.  Several protocol spaces have similar
requirements, so helper functions were placed in the Space class for
moving locally (using os.rename) and rsync.

Update store_aip to use the new functions.

Hwesta committed:
Staging path in default Space

refs #6248

Add staging path to default Space.  Only set the default transfer
source and AIP storage paths if they're unset.

Hwesta committed:
Rsync copies folders

Recursive option added to rsync command to copy folders properly.  Execute
permissions left alone, as directories that don't have execute bit don't
display permissions.

mistydemeo committed:
Package resource: expose under file name

Hwesta committed:
Extract file works for directories

Extract file requests will now check the filesystem to see if that path
exists in the package.  If not, and the package can be extracted, it will
fallback to the previous behaviour of extracting a file with the given path
and returning that.

Hwesta committed:
Fix moving folders locally

Bugs moving folders locally using os.rename or shutil.move, so changed to
always use move_rsync until properly fixed.

Hwesta committed:
post_move_from_storage_service takes set parameters

refs #6248

post_move_from_storage_service now expects the staging path (path relative
to the staging area for that Space), destination_path (path relative to the
Space.path for that Space) and package (optional, the Package that was
moved).

Hwesta committed:
Add migrations

refs #6466

Enable South, and update version.  Add migrations for 0.2.2 and 0.3.0
for locations and administration.

mistydemeo commented on storage_service/locations/models.py:
```diff
@@ -168,15 +172,160 @@ def browse(self, path):
+        logging.info("rsync command: {}".format(command))
+        try:
+            subprocess.check_call(command)
+        except Exception as e:
```
Catch `CalledProcessError` specifically?

mistydemeo commented on storage_service/locations/models.py:
```diff
@@ -168,15 +172,160 @@ def browse(self, path):
+    def _create_local_directory(self, path, mode=None):
+        """ Creates a local directory at 'path' with 'mode' (default 755). """
+        if mode is None:
+            mode = (stat.S_IRUSR + stat.S_IWUSR + stat.S_IXUSR +
```
Pass octal 0755 for brevity?

Hwesta replied:
I like the explicitness of the constants, though I'm convince-able on this one.

mistydemeo replied:
I've used too many Unix commands in my lifetime, so 0755 is actually more explicit/readable to me! Not a super strong opinion though.

mistydemeo commented on storage_service/locations/models.py:
```diff
@@ -168,15 +172,160 @@ def browse(self, path):
+                logging.warning("Could not create storage directory: {}".format(e))
+                raise
+
+        # Mode not getting set correctly in os.makedirs
```
Is this always true, or only in some circumstances?

Hwesta replied:
Seems to be true most of the time - `os.makedirs('foo/bar/baz', 0o777)` gives me directories with 755 permissions.  http://stackoverflow.com/questions/5231901/permission-problems-when-creating-a-dir-with-os-makedirs-python points out that the makedirs docs say the mode may be ignored.

mistydemeo replied:
Okay, maybe add more detail to the comment? It wasn't clear to me what this was guarding against.

Hwesta replied:
Would this be better? `os.makedirs may ignore the mode when creating directories, so force the permissions here. Some spaces (eg CIFS) doesn't allow chmod, so wrap it in a try-catch and ignore the failure.`

mistydemeo replied:
:thumbsup: Thank you!

mistydemeo commented on storage_service/locations/models.py:
```diff
@@ -168,15 +172,160 @@ def browse(self, path):
+        self.space._create_local_directory(destination_path)
+        return self.space._move_rsync(source_path, destination_path)
+
+    def move_from_storage_service(self, src_path, dest_path):
```
These two methods are doing very similar things, maybe the logic can be extracted?

Hwesta replied:
They are similar, but the path munging is different.  The similar logic is already extracted into `_create_local_directory` and `_move_rsync`.  Also, they may not always be identical (move_from used move_local instead of move_rsync for a while, and probably will again in future)

mistydemeo replied:
Okay, works for me.

mistydemeo commented on storage_service/locations/models.py:
```diff
@@ -236,24 +418,105 @@ class PipelineLocalFS(models.Model):
+        #     # FIXME dest_path is relative, and intended for staging_path, need
+        #     # real destination path - memoize something and retrieve it in
+        #     # move_from_storage_service?  Move to self.space.path/dest_path?
+        #     command = 'mkdir -p {dest_dir} && mv {src_path} {dest_path}'.format(
```
Quote `dest_dir` and `src_path`?

mistydemeo replied:
Uh, of course this isn't actual code just a comment. Ignore the noise.

Hwesta replied:
Given that may be copy-pasted into actual code, probably worthwhile to do anyway.

mistydemeo commented on storage_service/static/js/file-explorer.js:
```diff
@@ -199,6 +199,10 @@
 
+          // names are base64-encoded from the server; needs to be
+          // decoded for human reading
+          child.attributes.name = Base64.decode(child.attributes.name);
```
Is this reused elsewhere? Do you want to mutate `child`, or can you work from a copy of the name?

Hwesta replied:
This code is actually from #3, so I won't deal with this.

mistydemeo replied:
Oh no! The monster at the end of the PR... is _me_!

mistydemeo commented on storage_service/static/js/vendor/base64.js:
```diff
@@ -0,0 +1,216 @@
+/*
```
What is this branched from? This file already exists as `static/js/vendor/base64.js`.

Hwesta replied:
Originally branched from 0.x, then rebased onto qa/0.x when that became available.  Pull request is still against 0.x, so it has 4 extra commits relating to AICs and non-unicode/base64 stuff.

mistydemeo replied:
Ah, okay.

mistydemeo commented on storage_service/storage_service/urls.py:
```diff
@@ -43,7 +43,9 @@ def startup():
     space, space_created = locations_models.Space.objects.get_or_create(
-        access_protocol=locations_models.Space.LOCAL_FILESYSTEM, path='/')
+        access_protocol=locations_models.Space.LOCAL_FILESYSTEM,
+        path='/',
```
Still keeping `path='/'` when using `os.sep` below?

mistydemeo commented:
I only have nitpicks; this looks good! Seems like a good approach.

jhsimpson committed:
Adding uninstall script and supporting proper updating

Refs #6359 #6466

Ubuntu package will now work when updating from 0.2.2 to 0.3.0.
Will also work when uninstalling 0.3.0

Hwesta committed:
Add migration

Add migration for staging_path Space addition.

Hwesta commented:
Update with feedback from code review.  Also rebased off qa/0.x again, and got the migration changes, so added a migration.

mistydemeo commented:
Migrations! :D
